File: src/shared/exportTranscript.ts
// Transcript → a file the user can open or share: Markdown for reading, JSON for tooling. Pure and DOM-free so both
// tsconfigs (host and webview) can type-check it; the host formats, writes under exports/, and opens the result

import type { AgentBlock, ToolContent, Turn } from './transcript';

// Every user-facing label the export emits; English is the default so a host without a locale still produces a readable file
export interface ExportLabels {
  user: string;
  agent: string;
  project: string;
  exported: string;
  attachments: string;
  thinking: string;
  compacted: string;
  autoCompact: string;
  error: string;
}

export const EXPORT_LABELS_EN: ExportLabels = {
  user: 'User',
  agent: 'Agent',
  project: 'Project',
  exported: 'Exported',
  attachments: 'Attachments',
  thinking: 'Thinking',
  compacted: 'Context compacted',
  autoCompact: 'Automatic /compact',
  error: 'Error',
};

export interface ExportInput {
  title: string;
  agentName: string;
  cwd: string;
  exportedAt: string;
  turns: Turn[];
}

const TOOL_CONTENT_MAX = 4000;

// Resolves a blob name to its absolute store path so an agent image can export as a real file link;
// absent (a store without disk paths) the image degrades to a `[image]` note
export type ExportBlobPath = (name: string) => string | undefined;

export function exportMarkdown(input: ExportInput, labels: ExportLabels = EXPORT_LABELS_EN, blobPath?: ExportBlobPath): string {
  const parts: string[] = [
    `# ${input.title}`,
    `- **${labels.agent}**: ${input.agentName}\n- **${labels.project}**: ${input.cwd}\n- **${labels.exported}**: ${input.exportedAt}`,
    '---',
  ];
  for (const turn of input.turns) {
    const rendered = turn.role === 'user' ? userTurn(turn, labels) : agentTurn(turn, input.agentName, labels, blobPath);
    if (rendered) parts.push(rendered);
  }
  return `${parts.join('\n\n')}\n`;
}

function userTurn(turn: Extract<Turn, { role: 'user' }>, labels: ExportLabels): string {
  if (turn.auto) return `_${labels.autoCompact}_`;
  const out = [`### ${labels.user}`, '', turn.text];
  if (turn.attachments?.length) {
    const names = turn.attachments.map(a => a.kind === 'file' ? a.name : a.name ?? a.blob ?? 'image');
    out.push('', `> ${labels.attachments}: ${names.join(', ')}`);
  }
  return out.join('\n');
}

function agentTurn(turn: Extract<Turn, { role: 'agent' }>, agentName: string, labels: ExportLabels, blobPath?: ExportBlobPath): string {
  const out = [`### ${agentName}`];
  for (const b of turn.blocks) {
    const rendered = block(b, labels, blobPath);
    if (rendered) out.push('', rendered);
  }
  if (turn.stop === 'error' && turn.error) out.push('', `> ${labels.error}: ${turn.error.message}`);
  else if (turn.stop && turn.stop !== 'end_turn') out.push('', `> ${turn.stop}`);
  return out.join('\n');
}

function block(b: AgentBlock, labels: ExportLabels, blobPath?: ExportBlobPath): string | undefined {
  switch (b.type) {
    case 'thought':
      return `<details><summary>${labels.thinking}</summary>\n\n${b.text}\n\n</details>`;
    case 'text':
      return b.markdown;
    case 'image':
      return imageLine(b, blobPath);
    case 'plan':
      return b.entries.map(e => e.status === 'completed' ? `- [x] ${e.title}`
        : `- [ ] ${e.title}${e.status === 'in_progress' ? ' _(in progress)_' : ''}`).join('\n');
    case 'tool_call': {
      const head = `- **${b.verb}**${b.target ? ` \`${b.target}\`` : ''}${b.meta ? ` (${b.meta})` : ''}`;
      const content = (b.contents ?? (b.content ? [b.content] : [])).map(c => toolContent(c, blobPath)).filter(Boolean).join('\n');
      return content ? `${head}\n${content}` : head;
    }
    case 'permission':
      return undefined;
    case 'question': {
      if (!b.outcome) return undefined;
      return b.questions.map(q => {
        const a = b.answers?.[q.id];
        const answer = Array.isArray(a) ? a.join(', ') : a ?? '—';
        return `**Q:** ${q.text}\n\n**A:** ${answer}`;
      }).join('\n\n');
    }
    case 'compaction':
      return b.status === 'completed' ? `_${labels.compacted}_` : `_${labels.compacted}: ${b.status}_`;
    case 'plan_document':
      return `#### ${b.title}\n\n${b.markdown}`;
    default:
      return undefined;
  }
}

// An agent image exports as a link to its blob file on disk; uri-only ones keep the path the agent saved
function imageLine(i: { blob?: string; mimeType: string; uri?: string }, blobPath?: ExportBlobPath): string {
  const path = i.blob && blobPath?.(i.blob);
  if (path) return `![image](${path})`;
  if (i.uri) return `[image](${i.uri})`;
  return `[image: ${i.mimeType}]`;
}

function toolContent(c: ToolContent | undefined, blobPath?: ExportBlobPath): string | undefined {
  if (!c) return undefined;
  switch (c.type) {
    case 'text': {
      const body = c.text.length > TOOL_CONTENT_MAX ? `${c.text.slice(0, TOOL_CONTENT_MAX)}\n… (truncated)` : c.text;
      return `\`\`\`\n${body}\n\`\`\``;
    }
    case 'diff':
      // DiffLine.text already carries the + / - / space prefix; a hunk line's text is a localized note like "2 unchanged lines"
      return `\`\`\`diff\n${c.lines.map(l => l.kind === 'hunk' ? `@@ ${l.text} @@` : l.text).join('\n')}\n\`\`\``;
    case 'list':
      return c.items.map(i => `  - ${i}`).join('\n');
    case 'image':
      return imageLine(c, blobPath);
    default:
      return undefined;
  }
}

// `<slug>-<yyyyMMdd-HHmmss>.<md|json>`: title minus filesystem-hostile characters, whitespace collapsed to dashes,
// capped at 60 code points (CJK survives — only ASCII separators and control characters are dropped)
export function exportFileName(title: string, format: 'markdown' | 'json', now: Date): string {
  const slug = [...title
    .replace(/[/\\:*?"<>|\p{Cc}]/gu, '')
    .replace(/\s+/g, '-')
    .replace(/^[-.]+|[-.]+$/g, '')]
    .slice(0, 60).join('').replace(/[-.]+$/, '') || 'session';
  const p = (n: number) => String(n).padStart(2, '0');
  const stamp = `${now.getFullYear()}${p(now.getMonth() + 1)}${p(now.getDate())}-${p(now.getHours())}${p(now.getMinutes())}${p(now.getSeconds())}`;
  return `${slug}-${stamp}.${format === 'markdown' ? 'md' : 'json'}`;
}

File: rust/crates/acpira-shared/src/export_transcript.rs
//! Transcript → Markdown for reading (mirror of src/shared/exportTranscript.ts)

use crate::transcript::*;

pub struct ExportLabels {
  pub user: String,
  pub agent: String,
  pub project: String,
  pub exported: String,
  pub attachments: String,
  pub thinking: String,
  pub compacted: String,
  pub auto_compact: String,
  pub error: String,
}

pub struct ExportInput<'a> {
  pub title: &'a str,
  pub agent_name: &'a str,
  pub cwd: &'a str,
  pub exported_at: &'a str,
  pub turns: &'a [Turn],
}

const TOOL_CONTENT_MAX: usize = 4000;

fn slice16(s: &str, n: usize) -> String {
  let mut units = 0;
  let mut out = String::new();
  for c in s.chars() {
    units += c.len_utf16();
    if units > n {
      break;
    }
    out.push(c);
  }
  out
}

pub fn export_markdown(input: &ExportInput, labels: &ExportLabels, blob_path: &dyn Fn(&str) -> Option<String>) -> String {
  let mut parts = vec![
    format!("# {}", input.title),
    format!(
      "- **{}**: {}\n- **{}**: {}\n- **{}**: {}",
      labels.agent, input.agent_name, labels.project, input.cwd, labels.exported, input.exported_at
    ),
    "---".to_owned(),
  ];
  for turn in input.turns {
    let rendered = match turn {
      Turn::User(u) => user_turn(u, labels),
      Turn::Agent(a) => agent_turn(a, input.agent_name, labels, blob_path),
    };
    if !rendered.is_empty() {
      parts.push(rendered);
    }
  }
  format!("{}\n", parts.join("\n\n"))
}

fn user_turn(turn: &UserTurn, labels: &ExportLabels) -> String {
  if turn.auto == Some(true) {
    return format!("_{}_", labels.auto_compact);
  }
  let mut out = vec![format!("### {}", labels.user), String::new(), turn.text.clone()];
  if let Some(a) = turn.attachments.as_ref().filter(|a| !a.is_empty()) {
    let names: Vec<String> = a
      .iter()
      .map(|x| match x {
        Attachment::File { name, .. } => name.clone(),
        Attachment::Image { name, blob, .. } => name.clone().or_else(|| blob.clone()).unwrap_or_else(|| "image".into()),
        Attachment::Text { name, .. } => name.clone(),
      })
      .collect();
    out.push(String::new());
    out.push(format!("> {}: {}", labels.attachments, names.join(", ")));
  }
  out.join("\n")
}

fn stop_name(s: TurnStop) -> String {
  serde_json::to_value(s).ok().and_then(|v| v.as_str().map(str::to_owned)).unwrap_or_default()
}

fn agent_turn(turn: &AgentTurn, agent_name: &str, labels: &ExportLabels, blob_path: &dyn Fn(&str) -> Option<String>) -> String {
  let mut out = vec![format!("### {agent_name}")];
  for b in &turn.blocks {
    if let Some(r) = block(b, labels, blob_path) {
      out.push(String::new());
      out.push(r);
    }
  }
  match (&turn.stop, &turn.error) {
    (Some(TurnStop::Error), Some(e)) => {
      out.push(String::new());
      out.push(format!("> {}: {}", labels.error, e.message));
    }
    (Some(s), _) if *s != TurnStop::EndTurn => {
      out.push(String::new());
      out.push(format!("> {}", stop_name(*s)));
    }
    _ => {}
  }
  out.join("\n")
}

fn block(b: &AgentBlock, labels: &ExportLabels, blob_path: &dyn Fn(&str) -> Option<String>) -> Option<String> {
  Some(match b {
    AgentBlock::Thought(t) => format!("<details><summary>{}</summary>\n\n{}\n\n</details>", labels.thinking, t.text),
    AgentBlock::Text(t) => t.markdown.clone(),
    AgentBlock::Image(i) => image_line(i.blob.as_deref(), &i.mime_type, i.uri.as_deref(), blob_path),
    AgentBlock::Plan(p) => p
      .entries
      .iter()
      .map(|e| match e.status {
        PlanStatus::Completed => format!("- [x] {}", e.title),
        PlanStatus::InProgress => format!("- [ ] {} _(in progress)_", e.title),
        PlanStatus::Pending => format!("- [ ] {}", e.title),
      })
      .collect::<Vec<_>>()
      .join("\n"),
    AgentBlock::ToolCall(t) => {
      let head = format!(
        "- **{}**{}{}",
        t.verb,
        t.target.as_ref().filter(|x| !x.is_empty()).map(|x| format!(" `{x}`")).unwrap_or_default(),
        t.meta.as_ref().filter(|x| !x.is_empty()).map(|x| format!(" ({x})")).unwrap_or_default()
      );
      let items: Vec<&ToolContent> = match (&t.contents, &t.content) {
        (Some(l), _) => l.iter().collect(),
        (None, Some(c)) => vec![c],
        _ => vec![],
      };
      let content: Vec<String> = items.into_iter().filter_map(|c| tool_content(c, blob_path)).filter(|s| !s.is_empty()).collect();
      if content.is_empty() { head } else { format!("{head}\n{}", content.join("\n")) }
    }
    AgentBlock::Permission(_) => return None,
    AgentBlock::Question(q) => {
      q.outcome?;
      q.questions
        .iter()
        .map(|x| {
          let answer = match q.answers.as_ref().and_then(|a| a.get(&x.id)) {
            Some(serde_json::Value::Array(items)) => items.iter().filter_map(|v| v.as_str()).collect::<Vec<_>>().join(", "),
            Some(serde_json::Value::String(s)) => s.clone(),
            _ => "—".into(),
          };
          format!("**Q:** {}\n\n**A:** {answer}", x.text)
        })
        .collect::<Vec<_>>()
        .join("\n\n")
    }
    AgentBlock::Compaction(c) => {
      if c.status == CompactionStatus::Completed {
        format!("_{}_", labels.compacted)
      } else {
        let st = serde_json::to_value(c.status).ok().and_then(|v| v.as_str().map(str::to_owned)).unwrap_or_default();
        format!("_{}: {st}_", labels.compacted)
      }
    }
    AgentBlock::PlanDocument(p) => format!("#### {}\n\n{}", p.title, p.markdown),
    AgentBlock::Notice(n) => format!(
      "> {}: {}{}",
      if n.severity == Severity::Error { labels.error.as_str() } else { "Notice" },
      n.title,
      n.details.as_ref().filter(|d| !d.is_empty()).map(|d| format!("\n> {d}")).unwrap_or_default()
    ),
  })
}

fn image_line(blob: Option<&str>, mime: &str, uri: Option<&str>, blob_path: &dyn Fn(&str) -> Option<String>) -> String {
  if let Some(path) = blob.filter(|b| !b.is_empty()).and_then(blob_path) {
    return format!("![image]({path})");
  }
  match uri.filter(|u| !u.is_empty()) {
    Some(u) => format!("[image]({u})"),
    None => format!("[image: {mime}]"),
  }
}

fn tool_content(c: &ToolContent, blob_path: &dyn Fn(&str) -> Option<String>) -> Option<String> {
  Some(match c {
    ToolContent::Text { text } => {
      let body = if text.encode_utf16().count() > TOOL_CONTENT_MAX {
        format!("{}\n… (truncated)", slice16(text, TOOL_CONTENT_MAX))
      } else {
        text.clone()
      };
      format!("```\n{body}\n```")
    }
    ToolContent::Diff { lines, .. } => {
      let body: Vec<String> =
        lines.iter().map(|l| if l.kind == DiffKind::Hunk { format!("@@ {} @@", l.text) } else { l.text.clone() }).collect();
      format!("```diff\n{}\n```", body.join("\n"))
    }
    ToolContent::List { items } => items.iter().map(|i| format!("  - {i}")).collect::<Vec<_>>().join("\n"),
    ToolContent::Image(i) => image_line(i.blob.as_deref(), &i.mime_type, i.uri.as_deref(), blob_path),
  })
}

/// `<slug>-<yyyyMMdd-HHmmss>.<md|json>`; `stamp` is the local-time part, formatted by the host
pub fn export_file_name(title: &str, markdown: bool, stamp: &str) -> String {
  let cleaned: String = title.chars().filter(|c| !"/\\:*?\"<>|".contains(*c) && !c.is_control()).collect();
  let dashed = cleaned.split_whitespace().collect::<Vec<_>>().join("-");
  let mut dashed = dashed;
  // A leading / trailing whitespace run collapsed to a dash in the TS; split_whitespace drops it, then trim separators
  let trimmed = dashed.trim_matches(|c| c == '-' || c == '.').to_owned();
  dashed = trimmed;
  let capped: String = dashed.chars().take(60).collect();
  let slug = capped.trim_end_matches(['-', '.']);
  let slug = if slug.is_empty() { "session" } else { slug };
  format!("{slug}-{stamp}.{}", if markdown { "md" } else { "json" })
}
